### ISSUE TITLE
Lazy requests

### ISSUE THREAD

**Issue by azmeuk** [migraf/fhir-kindling#56]
I would like to retrieve a very large number of organizations (>100k) from a FHIR API. My understanding of this piece of code is that, when `FHIRQuery.all()` is called, every item is being requested on the API.

https://github.com/migraf/fhir-kindling/blob/999a394e04d6783870d7d0b11f10a4b803acbe18/fhir_kindling/fhir_query/query_response.py#L294-L303

This can lead to very long process time with no indication on what is going on on the client side (why is it taking so long? is the network stuck? is the dataset very big? is there an issue on the server side?)

I would suggest implementing, instead or in addition, a lazy behavior for requests where data are requested on access. That would mean, for instance if page size is 10, load the first page when the first item is accessed, and load the second pages only when the 11th item is accessed.
I suppose implementing this with `async` could make the waiting time even more transparent.

What do you think?

**Comment by migraf** [comment]
You are right and this shows me that some more comments outside of doc strings are required 😅.

Generally I have found it difficult to achieve a nice way of handling the response pagination of different FHIR servers i.e. some return the total number of resources matching the query and some don't. 

So i made the initial choice to fully resolve the pagination when `.all()` is used. This was also because I wanted to achieve easy access to the resources matching the query as a response property. But then calling `query_response.resources` would again lead to resolving the pagination from the server.

I have always been sort of uneasy about this though, since as you said it takes a long time for more than a few k resources and I agree that it would be nice to have lazy loading. Initial thoughts would be to add `lazy` and `count` parameters to the `all()` and `limit()` methods.

I agree that implementing this using `async` would be the way to go and `async` implementations of the public api of the `FhirServer` class have been on my todo list for a while.

Would you like to work on this? I'm pretty swamped with work at the moment, so it'll most likely be a 3-4 weeks until I can put more efforts into this library again.

**Comment by azmeuk** [comment]
> Would you like to work on this? I'm pretty swamped with work at the moment, so it'll most likely be a 3-4 weeks until I can put more efforts into this library again.

In a short term I won't probably have time to spend, but that may change in the future.

**Comment by migraf** [comment]
I think I might add some sort of optional feedback for, like a tqdm progress bar,  as an intermediate "fix"

**Comment by azmeuk** [comment]
An intermediate solution can be to pass a callback function that can be executed at each iteration. This way users can trigger custom code on each received item. Thus, you don't have to add dependencies to tqdm, and force users to output things on their console.

**Comment by migraf** [comment]
That also seems like a good idea. I´m thinking of an api something like this:

```python
n_per_page = 50
def callback(resources: List[FHIRAbstractModel):
    print(f" {len(resources)} resources received from the server")
    if len(resources) < n_per_page:
        print("Done")

query.all(lazy=True, callback=callback, count=n_per_page)
```

**Comment by azmeuk** [comment]
If `query.all` is lazy, then it will probably return a generator, and thus it should return almost immediately. It is quite easy to track the progression of the iteration of a generator, so I am not sure `callback` would be needed if `lazy` is implemented.

What do you think?

**Comment by migraf** [comment]
I agree. There's only a callback for now. I might add lazy functionality using a generator soon, but I might also include this with the async implementation.
Callbacks implemented in #62
